### PR TITLE
fixes #4668

### LINKED ISSUE #4668: type alias for generic typeclass doesn't work
The following doesn't compile on 0.14.2 or devel.

``` nimrod
type
  FooObj*[T] = object
    v*: T
  Foo1*[T] = FooObj[T]
  Foo2* = FooObj

proc foo1(x: Foo1) = echo "foo1"
proc foo2(x: Foo2) = echo "foo2"

var x: FooObj[float]
foo1(x)  # works
foo2(x)
# type mismatch: got (FooObj[system.float])
# but expected one of:
# proc foo2[Foo2](x: Foo2)
```

### PR DESCRIPTION
**Pull request by jcosborn**
This fixes #4668 but is potentially a breaking change since it modifies the type structures to make ``type Foo = Bar`` definitions get a tyAlias type.  This seemed like the correct thing to do.  If it is acceptable, I will try to fix the few regressions.  If not, I can instead fix this by hacking typeRel.

### CODE REVIEW

**Comment by Araq** [comment]
Note: The CIs appear to have **not** checked this.

**Comment by jcosborn** [comment]
I found another way which seemed much easier.

**Comment by krux02** [comment]
@jcosborn what do you mean with you found another way that is much easier?

**Comment by jcosborn** [comment]
Making ``Foo`` in ``type Foo = Bar`` have type ``tyAlias`` breaks the compiler in many other places which would take some time to fix.  Instead I just made that type declaration for a generic type ``Bar`` act as if it was written as ``type Foo[T] = Bar[T]`` (or whatever generic arguments it should have).  This already works properly in the rest of the compiler.  The only possibly down side is you can't distinguish these declarations from their types, but I'm not sure that is necessary.

**Comment by krux02** [comment]
@jcosborn Sorry I am a bit late with my reply. I don't fully understand your `type Foo[T] = Bar[T]` vs `type Foo = Bar` example. Dose Bar have a generic argument? Should Foo also have a generic argument? I have several aliases in my code like `type Vec4[T] = Vec[4,T]; Vec4f = Vec4[float32]` and I need to be able to resolve them as alias types to `Vec` via the macros api to compile them to glsl types.

**Comment by zah** [comment]
The approach taken here seems reasonable to me.

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -1106,6 +1106,22 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
         # type aliases are hard:
         var t = semTypeNode(c, x, nil)
         assert t != nil
+        if t.kind == tyGenericBody:
+          var tgi = newTypeS(tyGenericInvocation, c)
```
Comment (Araq) [inline compiler/semstmts.nim]: At the very least move it to a helper proc like ``repairGenericAliases``.

Comment (zah) [reply compiler/semstmts.nim]: There is some very similar code in `liftParamType` (defined in `semtypes.nim`).
I think you can extract a shared helper proc.

**Pull request closed without merging**